File: QEasyWidgets/Common/QWorker.py
from typing import Union, Optional, overload
from PyEasyUtils import singledispatchmethod, getNamesFromMethod, getClassFromMethod
from PySide6.QtCore import QObject, Signal, QThreadPool, QRunnable
from PySide6.QtWidgets import *

##############################################################################################################################

class WorkerSignals(QObject):
    """
    """
    started = Signal()
    finished = Signal()

    error = Signal(Exception)
    result = Signal(object)


class Worker(QRunnable):
    """
    """
    # @singledispatchmethod
    def __init__(self, autoDelete: bool = True):
        super().__init__()

        self.setAutoDelete(autoDelete)

        self.signals = WorkerSignals()

    # @__init__.register
    # def _(self, fn, *args, **kwargs):
    #     self.__init__()
    #     self.setTask(fn, *args, **kwargs)

    def setTask(self, fn, *args, **kwargs):
        self.fn = fn
        self.args = args
        self.kwargs = kwargs

    def run(self):
        self.signals.started.emit()

        try:
            result = self.fn(*self.args, **self.kwargs)
        except Exception as e:
            self.signals.error.emit(e)
        else:
            self.signals.result.emit(result)

        self.signals.finished.emit()

##############################################################################################################################

class WorkerManager:
    """
    """
    def __init__(self,
        executeMethod: object = ...,
        terminateMethod: Optional[object] = None,
        autoDelete: bool = True,
        threadPool: Optional[QThreadPool] = None
    ):
        self.threadPool = threadPool or QThreadPool()

        self.worker = Worker(autoDelete)

        executeClassName, executeMethodName = getNamesFromMethod(executeMethod)
        if executeClassName is not None:
            try:
                self.executeClassInstance = getClassFromMethod(executeMethod)()
                self.executeClassInstanceMethod = getattr(self.executeClassInstance, executeMethodName)
            except:
                self.executeClassInstanceMethod = executeMethod
        else:
            self.executeClassInstanceMethod = executeMethod

        if terminateMethod is None:
            self.terminateClassInstanceMethod = None
            return
        terminateClassName, terminateMethodName = getNamesFromMethod(terminateMethod)
        if terminateClassName is not None:
            try:
                self.terminateClassInstance = self.executeClassInstance if terminateClassName == executeClassName else getClassFromMethod(terminateMethod)()
                self.terminateClassInstanceMethod = getattr(self.terminateClassInstance, terminateMethodName)
            except:
                self.terminateClassInstanceMethod = terminateMethod
        else:
            self.terminateClassInstanceMethod = terminateMethod

    def execute(self, *executeParams):
        self.worker.setTask(self.executeClassInstanceMethod, *executeParams)
        self.threadPool.start(self.worker)

    def terminate(self, *terminateParams):
        if self.terminateClassInstanceMethod is None:
            return
        self.terminateClassInstanceMethod(*terminateParams)

##############################################################################################################################

File: c++/QEasyWidgets/Common/QWorker.cpp
#include "QWorker.h"

#include <QDebug>
#include <QException>


WorkerSignals::WorkerSignals(QObject *parent)
    : QObject(parent)
{
}


Worker::Worker(bool autoDelete)
    : m_signals(new WorkerSignals())
    , m_hasResult(false)
{
    setAutoDelete(autoDelete);
}


void Worker::setTask(std::function<QVariant()> task) {
    m_taskWithResult = task;
    m_hasResult = true;
}


void Worker::setTask(std::function<void()> task) {
    m_taskWithoutResult = task;
    m_hasResult = false;
}


void Worker::run() {
    emit m_signals->started();

    try {
        if (m_hasResult && m_taskWithResult) {
            QVariant result = m_taskWithResult();
            emit m_signals->result(result);
        } else if (!m_hasResult && m_taskWithoutResult) {
            m_taskWithoutResult();
        }
    } catch (const std::exception &e) {
        emit m_signals->error(QString::fromStdString(e.what()));
    } catch (...) {
        emit m_signals->error("Unknown error occurred");
    }

    emit m_signals->finished();
}


WorkerManager::WorkerManager(QThreadPool *threadPool, QObject *parent)
    : QObject(parent)
    , m_threadPool(threadPool ? threadPool : QThreadPool::globalInstance())
    , m_worker(nullptr)
    , m_terminateWorker(nullptr)
{
}


WorkerManager::~WorkerManager() {
    if (m_worker) {
        delete m_worker;
    }
    if (m_terminateWorker) {
        delete m_terminateWorker;
    }
}


void WorkerManager::setExecuteTask(std::function<QVariant()> task) {
    m_executeTask = task;
}


void WorkerManager::setTerminateTask(std::function<void()> task) {
    m_terminateTask = task;
}


void WorkerManager::execute() {
    if (!m_executeTask) return;

    if (m_worker) {
        delete m_worker;
    }

    m_worker = new Worker(true);
    m_worker->setTask(m_executeTask);
    m_threadPool->start(m_worker);
}


void WorkerManager::terminate() {
    if (!m_terminateTask) return;

    if (m_terminateWorker) {
        delete m_terminateWorker;
    }

    m_terminateWorker = new Worker(true);
    m_terminateWorker->setTask(m_terminateTask);
    m_threadPool->start(m_terminateWorker);
}